src/USER-BOCS/fix_bocs.h

```diff
@@ -151,7 +151,8 @@ class FixBocs : public Fix {
   void nhc_press_integrate();
 
   int read_F_table(char *, int);
-  void build_cubic_splines(double **);
+  int build_linear_splines(std::vector<std::vector<double>>);
```

Comment:
This API change is not good. Unless is is a (small) temporary array, multi-dimensional data should be allocated with `memory->create()` and deleted with `memory->destroy()`. But this is even worse since C++ does call by value and thus you create a needless copy and cannot pass it back. If the data is not used elsewhere, it should not be passed as an argument, but allocated locally, if it is needed elsewhere, it cannot be passed by value.

Reply:
Good catch on the function parameter. Thank you.  I've been spending so much time in the languages that pass arrays and objects by reference automatically that I let my guard down. But the pass-by-value problem is easily fixed by making the parameter a const reference to the vector. The vector is quite local; it exists only during the course of reading and processing an input data file. The data structure passed to the compute module is still a `double **`.

Checking for understanding here: Do you have a more general concern about the ability of vectors to handle reasonably large multidimensional arrays of doubles? The coder overhead is high to correctly manage the calloc/free code that was being used in this code before. A switch to vectors decreases that burden without having to change any of the code that accesses the actual array elements. Now that we have C++11 at our command, are there good reasons not to use vectors consistently for local data structures?

Reply:
I've submitted a change to the build_xxx_splines signatures to switch to passing a const reference to the vector.

Reply:
> Checking for understanding here: Do you have a more general concern about the ability of vectors to handle reasonably large multidimensional arrays of doubles? The coder overhead is high to correctly manage the calloc/free code that was being used in this code before. A switch to vectors decreases that burden without having to change any of the code that accesses the actual array elements. Now that we have C++11 at our command, are there good reasons not to use vectors consistently for local data structures?

the problem of the previous code is that it didn't use the memory class, which can allocate and delete multi-dimensional arrays in one go *and* nullify the pointer after deleting. That made it look clumsy. Also, the memory class has provisions for allocating memory with a preferred alignment and is integrated into the error reporting when malloc fails. When using C++ features you would also have to wrap everything into try/catch blocks to cleanly handle memory allocation issues and shut down parallel runs cleanly. My take on using containers is: we use them only when we can take advantage of their additional features, e.g. use iterators, or the size() member function (in case the size is unknown) or want to add/remove elements incrementally. What to use also depends on the specific use case. You can see that in the use of the tokenizer class (for speed) or the utils::split_words() function (for consistency and convenience).

Reply:
Crikey. @akohlmey, I think I'm finally understanding your first comment above from 4 Aug. Sorry to seem dense. The USER/fix_bocs code is the only code within LAMMPS I've ever worked on. I simply took it on its own terms and tried to fix specific problems from there. It contained no use of the Memory class, only raw calloc and free calls. 

Not taking anything for granted here, would you please confirm the accepted best practice for replacing calloc, which both allocates memory and initializes all bits to 0 (where the intent was probably to initialize all entries in the array to 0.0).

If the old code was:
`double *z = (double *)calloc(n, sizeof(double));`
...
`free(z);`

The LAMMPS best practice code would be:
`double *z;`
`memory->create(z, n, "z");`
`for (int i = 0; i < n; i++) { z[i] = 0.0; }`
...
`memory->destroy(z);`

Reply:
@eagunn yes, that works. But I don't think that the clearing of the array is needed. Also the whole file reading code looks needlessly convoluted and complex. Like it was taken from some other application. On top of that you need to keep in mind, that many people that contribute to scientific software have no formal training in programming and sometimes get stuck in an "expert beginner" state, where their programming is full of quirks that are typical for beginners, but have become experienced enough in their ways to more-or-less achieve their objectives. When looking at the example folder, this looks like the clearing of the allocated storage is not needed, since the array is dimensioned for the data to be read.
I would suggest to look at pair style table as a template for how to read such tables.

You should also need to keep in mind, that you are looking at a "user" package and particularly one that was added at a time, when the barrier for adding code was deliberately kept very low. This code would not be merged like this anymore. We now work with people to improve their code to meet higher minimum standards, as we have found out that this is more effective that to merge it first and then try to rectify style issues later. This also reflects that change that I am not handling the first contact with contributors, while in the past I would often only see new contributions after they were already added and could give my suggestions and comments only later on (which then frequently resulted in me doing the touching up of the code). So we have more conventional protocols now (even though I - as well - have no formal training, but had lots of guidance from people who had and also have now 25+ years of working on maintaining scientific software. so if there is a mistake to be made, I have made it myself and thus can more easily spot it when other untrained people do them.).

Reply:
I've converted all dynamically allocated arrays/matrices of doubles to use LAMMPS-standard memory class.